src/Build.UnitTests/WarningsAsMessagesAndErrors_Tests.cs

```diff
@@ -310,6 +391,77 @@ public void WarningsAsErrorsAndMessages_Tests(string WarningsAsMessages,
             }
         }
 
+        [Theory]
+
+        [InlineData(true)]// Log MSB1234, BuildWarningsNotAsErrors takes priority
+        [InlineData(false)]
+        public void WarningsNotAsErrorsAndMessages_Tests(bool useMSPrefix)
+        {
+            string Warning = "MSB1235";
+            using (TestEnvironment env = TestEnvironment.Create(_output))
+            {
+                string prefix = useMSPrefix ? "MSBuild" : "";
+                TransientTestProjectWithFiles proj = env.CreateTestProjectWithFiles($@"
+                <Project>
+                    <PropertyGroup>
+                        <{prefix}TreatWarningsAsErrors>true</{prefix}TreatWarningsAsErrors>
+                        <{prefix}WarningsNotAsErrors>{Warning}</{prefix}WarningsNotAsErrors>
+                    </PropertyGroup>
+                    <Target Name='Build'>
+                        <Warning Text=""some random text"" Code='{Warning}' />
+                    </Target>
+                </Project>");
+
+                MockLogger logger = proj.BuildProjectExpectSuccess();
+
+                logger.WarningCount.ShouldBe(1);
+                logger.ErrorCount.ShouldBe(0);
+
+                logger.AssertLogContains(Warning);
+            }
+        }
+
+
+
+        [Theory]
+        [InlineData("TreatWarningsAsErrors", "true", false)] // All warnings are treated as errors
+        [InlineData("WarningsAsErrors", "MSB1007", false)]
+        [InlineData("WarningsAsMessages", "MSB1007", false)]
+        [InlineData("WarningsNotAsErrors", "MSB1007", true)]
+        public void WarningsChangeWaveTest(string property, string propertyData, bool treatWarningsAsErrors)
+        {
+            using (TestEnvironment env = TestEnvironment.Create(_output))
+            {
+                string warningCode = "MSB1007";
+                string treatWarningsAsErrorsCodeProperty = treatWarningsAsErrors ? "<MSBuildTreatWarningsAsErrors>true</MSBuildTreatWarningsAsErrors>" : "";
+                env.SetEnvironmentVariable("MSBUILDDISABLEFEATURESFROMVERSION", ChangeWaves.Wave17_14.ToString());
+                TransientTestProjectWithFiles proj = env.CreateTestProjectWithFiles($@"
+                <Project>
+                    <PropertyGroup>
+                        {treatWarningsAsErrorsCodeProperty}
+                        <{property}>{propertyData}</{property}>
+                    </PropertyGroup>
+                    <Target Name='Build'>
+                        <Warning Text=""some random text"" Code='{warningCode}' />
+                    </Target>
+                </Project>");
+                if (treatWarningsAsErrors)
+                {
+                    // Since the "no prefix" variations can't do anything with the change wave disabled, this should always fail.
+                    MockLogger logger = proj.BuildProjectExpectFailure();
```

Comment:
Mhm. Asserting on a specific error would require me digging quite deep into the proj.BuildProjectExpectFailure(); harness which is somewhat out of scope although I think that for the future we might want to consider refactoring all of these tests since some parts of it are clunky at best.
I've added checking for the error in the logging. If we fail with an unrelated error (e.g. inside of the handling), it should fail ungracefully and not log anything, failing the test.
Furthermore, I added a positive test - WarningsNotAsErrors should work as no-op if MSBuildTreatWarningsAsErrors is not set, e.g. it should succeed without doing anything. (or more specifically, be treated as any other unreserved property)

Reply:
The added assertion on errors count and content of the logger is exactly what I had in mind! :-)
It might be slightly more safe to verify that the `warningCode` is really an error - something like:

```
logger.AssertLogContains($"Error: {warningCode}");
```

or if it needs a pattern match then something like (would need to be bit more elaboret to ensure singleline):

```
logger.FullLog.ShouldMatch($"Error: * {warningCode}")
```


But other than that looks good - feel free to reflect or not, resolve this and merge ;-)